config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java

```diff
@@ -0,0 +1,95 @@
+        for (int i = this.children.size() - 1; i >= 0; i--) {
+            ConfigNode currNode = this.children.get(i);
+            if (currNode instanceof ConfigNodeKeyValue) {
+                keyValueIndexes.add(i);
+            }
+        }
+    }
+
+    public ArrayList<ConfigNode> children() {
+        return children;
+    }
+
+    public String render() {
+        StringBuilder renderedText = new StringBuilder();
+        for (ConfigNode child : children) {
+            renderedText.append(child.render());
+        }
+        return renderedText.toString();
+    }
+
+    protected ConfigNodeComplexValue changeValueOnPath(Path desiredPath, ConfigNodeValue value) {
+        ArrayList<ConfigNode> childrenCopy = (ArrayList<ConfigNode>)(children.clone());
+        boolean replaced = value == null;
+        ConfigNodeKeyValue node;
+        Path key;
+        for (Integer keyValIndex : keyValueIndexes) {
+            node = (ConfigNodeKeyValue)children.get(keyValIndex.intValue());
+            key = Path.newPath(node.key().render());
+            if (key.equals(desiredPath)) {
+                if (!replaced) {
+                    childrenCopy.set(keyValIndex.intValue(), node.replaceValue(value));
+                    replaced = true;
+                }
+                else
+                    childrenCopy.remove(keyValIndex.intValue());
+            } else if (desiredPath.startsWith(key)) {
+                if (node.value() instanceof ConfigNodeComplexValue) {
+                    Path remainingPath = desiredPath.subPath(key.length());
+                    if (!replaced) {
+                        node = node.replaceValue(((ConfigNodeComplexValue) node.value()).changeValueOnPath(remainingPath, value));
+                        if (node.render() != children.get(keyValIndex.intValue()).render())
+                            replaced = true;
+                        childrenCopy.set(keyValIndex.intValue(), node);
+                    } else {
+                        node = node.replaceValue(((ConfigNodeComplexValue) node.value()).removeValueOnPath(remainingPath));
+                        childrenCopy.set(keyValIndex.intValue(), node);
+                    }
+                }
+            }
+        }
+        return new ConfigNodeComplexValue(childrenCopy);
+    }
+
+    public ConfigNodeComplexValue setValueOnPath(Path desiredPath, ConfigNodeValue value) {
+        ConfigNodeComplexValue node = changeValueOnPath(desiredPath, value);
+
+        // If the desired Path did not exist, add it
+        if (node.render().equals(render())) {
+            ArrayList<ConfigNode> childrenCopy = (ArrayList<ConfigNode>)children.clone();
+            ArrayList<ConfigNode> newNodes = new ArrayList<ConfigNode>();
+            newNodes.add(new ConfigNodeBasic(Tokens.newLine(null)));
+            newNodes.add(new ConfigNodeKey(Tokens.newUnquotedText(null, desiredPath.render())));
```

Comment:
It should be pretty easy to take the `ConfigNodeKey` class and make it multi-token, so I'll get to work on that. I think we can easily keep track of the tokens used to parse a path in a Collection, then pass that into the `PathBuilder` so that it saves them and can thus reproduce the original tokens/text.